Add basic ExchangeMailboxUser model which deals with #171

grantholle opened the pull request:
I had to change the underlying `applyObjectClassScopes` to account for star queries. I'm still tinkering with different objects, but this worked with my organization to retrieve "mailbox users" aka users with a mailbox guid. Let me know what you think. 

I'm working on ExchangeServer and ExchangeDatabase models now as well, but the objectclass that ldaptools references has aren't returning results for me. I _do_ retrieve results for our servers under the `Computer` model, but can't pinpoint anything that distinguishes it as an "Exchange Server" that could be universally true. Google is failing me also.

stevebauman commented on src/Models/Model.php:
```diff
@@ -412,8 +412,11 @@ public function registerGlobalScopes($builder)
+        foreach (static::$objectClasses as $key => $objectClass) {
+            $attribute = is_string($key) ? $key : 'objectclass';
+
+            echo "{$attribute} = {$objectClass}\n";
```
Think this might be an accidental addition

grantholle replied:
You bet! haha

stevebauman commented:
Do Exchange Mailboxes need to inherit all the functionality of the base `User` model?

Do they have passwords, belong to groups, and have the same `$dates`, for example?

grantholle commented:
Yes, they basically just have some extra attributes that make them "have a mailbox" (like having a `msExchMailboxGuid` attribute). They are a traditional user, though.

stevebauman commented:
Ok -- I wonder if we should instead add this as a method scope to the `ActiveDirectoryBuilder` instead (Or maybe to the `User` model)?

For example:

https://github.com/DirectoryTree/LdapRecord/blob/master/src/Query/Model/ActiveDirectoryBuilder.php

```php
// src/Query/Model/ActiveDirectoryBuilder.php

public function whereHasMailbox()
{
    return $this->whereHas('msExchMailboxGuid');
}
```

Since the exchange mailbox doesn't have any unique object classes, I don't want to have to apply unique scoping logic using the `$objectClasses` property. The property should only contain object classes, not query parameters to limit results. Doing this will also impact the creation of new models in the directory as well, as we would have to filter out these extra object classes somehow.

Are you able to inspect the exchange mailboxes object classes in your AD and see what it contains?

grantholle commented:
Yes, the objectclasses for a "mailbox user" are just user ones. They don't receive a unique objectclass.

I like the idea of having a separate model, but either way is fine. I've reverted the objectclass logic and added a global scope to the `ExchangeMailboxUser` (still extending `User`). What do you think about that?

stevebauman commented:
Hi @grantholle, due to this PR, I've added the ability to add local model scopes that covers this use case, and have added the local scope `whereHasMailbox()`:

```php
$mailboxUsers = User::whereHasMailbox()->get();
```

I felt that a whole new class and global scope was a bit much for retrieving users who have a mailbox. I didn't want to add a new model when it wasn't actually it's own separate object in Active Directory.

Thanks a ton for your contribution and helping me understand the use case! I'm creating a new release as we speak with this built-in.

If you think there should be other "mailbox" oriented methods added to the `User` model, please let me know 👍

grantholle commented:
Yes, local scopes are great. Thanks!
